README.md
# IBMiNagios

## Features

- API over IBMi Web Service
- No additional IBMi user is needed
- Security
  - API access only from specified IP addresses
  - API access only with valid authorisation key
  - Read only
  - As SQL statement only SELECT is allowed
  - IFS access is allowed only for objects to which the user QTMHHTP1 has rights
- The Nagios check command is the script *ibminagios.sh*, this script calls the IBMi Web Service (IBMi program *ibminagios*) with authorisation key, command and command parameters and processes the output of the API.

## Configuration

- IBMi objects
  - /etc/ibminagios.conf
    - rights (only the user *QTMHHTP1* should able to read and write the file):
      ```
      -rw------- QTMHHTP1  0
      ```
    - content:
      ```
      AUTHKEY=halala
      ```
      - Authorisation key to use the API.
  - IBMNAGIOS/IBMNAGIOS *PGM
- Nagios objects
  - the tool *wget* must be installed
  - /etc/ibminagios.conf
    - rights (only the user *nagios* should able to read and write the file):
      ```
      -rw------- nagios nobody
      ```
    - content:
      ```
      <host>:AUTHKEY=halala
      <host>:HTTPPORT=81
      ```
      - Authorisation key and API port fpr each monitored hosts.
  - *&lt;nagios path&gt;*/etc/checkcommands.conf
    ```
    define command {
      command_name ibminagios
      command_line $USER1$/ibminagios.sh $HOSTADDRESS$ "$ARG1$" "$ARG2$" "$ARG3$" "$ARG4$" "$ARG5$" "$ARG6$" "$ARG7$" "$ARG8$" "$ARG9$"
    }
    ```
  - *&lt;nagios path&gt;*/libexec/ibminagios.sh
- IBMi Web Service
  - httpd.conf
    ```
    ScriptAlias /ibminagios /QSYS.LIB/IBMINAGIOS.LIB/IBMINAGIOS.PGM
    <Directory /QSYS.LIB/IBMINAGIOS.LIB/>
      Order Deny,Allow
      Allow From 10.43.85.172
      Deny From all
      SetHandler cgi-script
      Options +ExecCGI
      SetEnv IBMI_NAGIOS_CONF_FILE /etc/ibminagios.conf
    </Directory>
    ```
    - you should change the IP address to the address of the nagios (only nagios will be able use this API).
    - you may use the SetEnv IBMI_NAGIOS_CONF_FILE to define the path of the ibminagios.conf file on the IBMi. The default path is /etc/ibminagios.conf.

## ibminagios.sh

```
ibminagios.sh <host> <command>
  command: [ syssts aspused <warning_condition> <critical_condition> ] |
           [ job [ msgw <warning_condition> <critical_condition> <exception_of_job_names> ] |
                 [ act <warning_condition> <critical_condition> <job_name> ] ] |
           [ outq splfcount <warning_condition> <critical_condition> <outq_name> ] |
           [ sql - <warning_condition> <critical_condition> <sql_statement> ] |
           [ sbs [ act <warning_condition> <critical_condition> <sbs_name> ] |
                 [ jobcount <warning_condition> <critical_condition> <sbs_name> ] ] |
           [ msgq [ inq <warning_condition> <critical_condition> <msgq_name> <msgid | *ALL> <minutes> ] |
                  [ search <warning_condition> <critical_condition> <msgq_name> <msgid | *ALL> <minutes> [ <exception_of_message_ids> ] ] ] |
           [ ifs [ filecount <warning_condition> <critical_condition> <dir_name> ] |
                 [ dircount <warning_condition> <critical_condition> <dir_name> ] |
                 [ filesize <warning_condition> <critical_condition> <dir_name> <file_name> ] |
                 [ oldestfile <warning_condition> <critical_condition> <dir_name> ] ]
  warning_condition and critical_condition:
    format: [ <condition><value> | - ]
      -  : no condition processing
    conditions:
      lt : less than 
      le : less equal 
      eq : equal
      ne : not equal 
      ge : greater equal
      gt : greater than
    values (per command and subcommand):
      syssts aspused : percent of usage
      job msgw       : number of jobs
      job act        : number of jobs
      outq splfcount : number of spool files
      sql            : return value of the sql statement
      sbs act        : status of the subsystem
      sbs jobcount   : number of jobs
      msgq inq       : number of messages
      msgq search    : number of messages
      ifs filecount  : number of files
      ifs dircount   : number of directories
      ifs filesize   : size of the file in bytes
      ifs oldestfile : age of the file in seconds
    special value: <NULL>
  exception_of_job_names: Job names separated with comma
  exception_of_message_ids: Message ids separated with comma  
```

## ibminagios

- Usage
  ```
  https://<host>:<port>/ibminagios?authkey=<auth_key>&cmd=<command>
  ```
- Commands
  - 001
    - **Required param:** -
    - **Optional param:** -
    - **Output:** Info=|AspUsed=&lt;x&gt;|
  - 002
    - **Required param:** OUTQNAME
    - **Optional param:** OUTQLIB
    - **Output:** Info=|NumberOfFiles=&lt;x&gt;|
  - 003
    - **Required param:** -
    - **Optional param:** STATUS, JOBNAME
    - **Output:** Info=|JobName=&lt;x&gt;|JobUser=&lt;x&gt;|JobNumber=&lt;x&gt;|Status=&lt;x&gt;|Type=&lt;x&gt;|SubType=&lt;x&gt;|ActiveJobStatus=&lt;x&gt;|SubsystemName=&lt;x&gt;|
  - 004
    - **Required param:** SQL
    - **Optional param:** -
    - **Output:** Info=|Value=&lt;x&gt;|
  - 005
    - **Required param:** -
    - **Optional param:** SBSNAME
    - **Output:** Info=|SubsystemName=&lt;x&gt;|SubsystemLibraryName=&lt;x&gt;|Status=&lt;x&gt;|ActiveJobs=&lt;x&gt;|
  - 006
    - **Required param:** MSGQNAME, MINUTES
    - **Optional param:** MSGID
    - **Output:** Info=|MessageId=&lt;x&gt;|Date=&lt;x&gt;|Time=&lt;x&gt;|Severity=&lt;x&gt;|Type=&lt;x&gt;|
  - 007
    - **Required param:** DIR
    - **Optional param:** FILE
    - **Output:** Info=|Name=&lt;x&gt;|Type=&lt;x&gt;|Size=&lt;x&gt;|ChangedDateTime=&lt;x&gt;|ModifiedDateTime=&lt;x&gt;|AccessedDateTime=&lt;x&gt;|CodePage=&lt;x&gt;|

## Example services

- System ASP used
  Warning from 75%, critical from 90%.
  ```
  check_command ibminagios!syssts!aspused!ge75!ge90
  ```
- Number of spool files in output queue QEZDEBUG
  Warning if there is at least 1 spool file in the queue, no critical check.
  ```
  check_command ibminagios!outq!splfcount!ge1!-!QEZDEBUG
  ```
- Number of jobs with active status MSGW (with two exceptions)
  Warning if there is at least 1 job has active status MSGW, nor critical check. The job names USVJOB and RH_STRZX60 are exceptions (they have always active status MSGW).
  ```
  check_command ibminagios!job!msgw!ge1!-!USVJOB,RH_STRZX60
  ```
- Number of active jobs with name USVJOB
  No warning check, critical if the number of of jobs is 0 (the job is not active).
  ```
  check_command ibminagios!job!act!-!eq0!USVJOB
  ```
- Number of not active iCluster nodes
  Warning if the number of not active nodes greater equal 1, no critical check. 
  ```
  check_command ibminagios!sql!-!ge1!-!SELECT count(*) AS satz FROM icluster.dmnodes WHERE status <> '*ACTIVE'
  ```
- Number of jobs in the subsystem ZEITSBSD
  Warning if the number of jobs is less than 7, no critical check.
  ```
  check_command ibminagios!sbs!jobcount!lt7!-!ZEITSBSD
  ```
- Status of the OMK subsystem
  No warning check, critical if the status of the subsystem is not *ACTIVE
  ```
  check_command ibminagios!sbs!act!-!ne*ACTIVE!OMK
  ```
- Number of messages in the message QSECOFR in the last 15 minutes
  Warning if the number of messages in the last 15 minutes is greater than 0, no critical check.
  ```
  check_command ibminagios!msgq!search!gt0!-!QSECOFR!*ALL!15
  ```
- Number of files in the IFS directory /home/transfer
  Warning if the number of files is greater than 100, no critical check.
  ```
  check_command ibminagios!ifs!filecount!gt100!-!/home/transfer
  ```
- Oldest file in the IFS directory /home/transfer
  Warning of the age of the oldest file is greater than 32000000 (aproximately 1 year), no critical check.
  ```
  check_command ibminagios!ifs!oldestfile!gt32000000!-!/home/transfer/archive
  ```
- Size of the IFS file /tmp/sendmail.log
  Warning if the size of the file is greater than 4000000 (aproximately 4 MB), no critical check.
  ```
  check_command ibminagios!ifs!filesize!gt4000000!-!/tmp!sendmail.log
  ```


CHANGES.md
# Changes

## 2023.10.04

- bugfix ifs file name
- new shell subcommand
  - ifs emptyfilecount

## 2023.04.16

- new API command 008 (journal receivers)
- new shell command and subcommands
  - jrnrcv partialcount
  - jrnrcv oldestdetached

## 2022.06.24

- changing the parameter *msgq_name* for the command *msgq*
  - adding possibility to specify the library name, example: USERLIB/ERRORQ
  - if the library is not specified, *LIBL will be used (example: QSYSOPR)

## 2022.06.13

- adding the parameter *exception_of_message_ids* for the command *msgq* subcommand *search*

## 2022.06.03

- new API command 007 (IFS directory entries)
- new shell command and subcommands
  - ifs filesize
  - ifs filecount
  - ifs dircount
  - ifs oldestfile




src/ibminagios.sh
#!/bin/sh

#
# ibminagios.sh
#
# Created by: github.com/TickMogne, 2022.05.31
#

#
# usage()
#
# Parameters:
#  -
#
# Return value:
#  -
#
usage() {
  cat <<EOF
ibminagios.sh <host> <command>
  command: [ syssts aspused <warning_condition> <critical_condition> ] |
           [ job [ msgw <warning_condition> <critical_condition> <exception_of_job_names> ] |
                 [ act <warning_condition> <critical_condition> <job_name> ] ] |
           [ outq splfcount <warning_condition> <critical_condition> <outq_name> ] |
           [ sql - <warning_condition> <critical_condition> <sql_statement> ] |
           [ sbs [ act <warning_condition> <critical_condition> <sbs_name> ] |
                 [ jobcount <warning_condition> <critical_condition> <sbs_name> ] ] |
           [ msgq [ inq <warning_condition> <critical_condition> <msgq_name> <msgid | *ALL> <minutes> ] |
                  [ search <warning_condition> <critical_condition> <msgq_name> <msgid | *ALL> <minutes> [ <exception_of_message_ids> ] ] ] |
           [ ifs [ filecount <warning_condition> <critical_condition> <dir_name> ] |
                 [ dircount <warning_condition> <critical_condition> <dir_name> ] |
                 [ filesize <warning_condition> <critical_condition> <dir_name> <file_name> ] |
                 [ oldestfile <warning_condition> <critical_condition> <dir_name> ] |
                 [ emptyfilecount <warning_condition> <critical_condition> <dir_name> ] ]
           [ jrnrcv [ partialcount <warning_condition> <critical_condition> ] |
                    [ oldestdetached <warning_condition> <critical_condition> [ <journal> ] ] ]
  warning_condition and critical_condition:
    format: [ <condition><value> | - ]
      -  : no condition processing
    conditions:
      lt : less than 
      le : less equal 
      eq : equal
      ne : not equal 
      ge : greater equal
      gt : greater than
    values (per command and subcommand):
      syssts aspused        : percent of usage
      job msgw              : number of jobs
      job act               : number of jobs
      outq splfcount        : number of spool files
      sql                   : return value of the sql statement
      sbs act               : status of the subsystem
      sbs jobcount          : number of jobs
      msgq inq              : number of messages
      msgq search           : number of messages
      ifs filecount         : number of files
      ifs dircount          : number of directories
      ifs filesize          : size of the file in bytes
      ifs oldestfile        : age of the file in seconds
      jrnrcv partialcount   : number of journal receivers
      jrnrcv oldestdetached : age of the journal receiver (datached date and time) in seconds
    special value: <NULL>
  exception_of_job_names: Job names separated with comma
  exception_of_message_ids: Message ids separated with comma
EOF
}

#
# check_condition()
#
# Parameters:
#  1. condition
#  2. value to check
#
# Return value:
#  Returns 1 if the result is true, otherwise returns 0.
#
check_condition() {
  local A
  local T

  if [ "$1" = "-" ]; then
    return 0
  fi

  VALUE=`echo $1 | sed -e 's/^[a-z][a-z]\(.*\)/\1/'`
  T=`echo $VALUE | grep '^[0-9]*$'`
  if [ "$T" = "" ]; then
    T="STRING"
  else
    T="NUMBER"
  fi

  A=`echo $1 | grep '^lt'` 
  if [ "$A" != "" ]; then
    VALUE=`echo $1 | sed -e 's/^lt\(.*\)$/\1/'`
    case $T in
      NUMBER)
        if [ $2 -lt $VALUE ]; then
          return 1
        fi
        ;;
      STRING)
        if [[ "$2" < "$VALUE" ]]; then
          return 1
        fi
        ;;
    esac
    return 0
  fi

  A=`echo $1 | grep '^le'` 
  if [ "$A" != "" ]; then
    VALUE=`echo $1 | sed -e 's/^le\(.*\)$/\1/'`
    case $T in
      NUMBER)
        if [ $2 -le $VALUE ]; then
          return 1
        fi
        ;;
      STRING)
        if [[ "$VALUE" > "$2" ]]; then
          return 1
        fi
        ;;
    esac
    return 0
  fi

  A=`echo $1 | grep '^eq'` 
  if [ "$A" != "" ]; then
    VALUE=`echo $1 | sed -e 's/^eq\(.*\)$/\1/'`
    case $T in
      NUMBER)
        if [ $2 -eq $VALUE ]; then
          return 1
        fi
        ;;
      STRING)
        if [ "$2" = "$VALUE" ]; then
          return 1
        fi
        ;;
    esac
    return 0
  fi

  A=`echo $1 | grep '^ne'` 
  if [ "$A" != "" ]; then
    VALUE=`echo $1 | sed -e 's/^ne\(.*\)$/\1/'`
    case $T in
      NUMBER)
        if [ $2 -ne $VALUE ]; then
          return 1
        fi
        ;;
      STRING)
        if [ "$2" != "$VALUE" ]; then
          return 1
        fi
        ;;
    esac
    return 0
  fi

  A=`echo $1 | grep '^ge'` 
  if [ "$A" != "" ]; then
    VALUE=`echo $1 | sed -e 's/^ge\(.*\)$/\1/'`
    case $T in
      NUMBER)
        if [ $2 -ge $VALUE ]; then
          return 1
        fi
        ;;
      STRING)
        if [[ "$VALUE" < "$2" ]]; then
          return 1
        fi
        ;;
    esac
    return 0
  fi

  A=`echo $1 | grep '^gt'` 
  if [ "$A" != "" ]; then
    VALUE=`echo $1 | sed -e 's/^gt\(.*\)$/\1/'`
    case $T in
      NUMBER)
        if [ $2 -gt $VALUE ]; then
          return 1
        fi
        ;;
      STRING)
        if [ "$2" > "$VALUE" ]; then
          return 1
        fi
        ;;
    esac
    return 0
  fi

  return 0
}

#
# check_conditions()
#
# Parameters:
#  1. critical condition
#  2. warning condition
#  3. value to check
#
# Return value:
#  Returns the nagios status (OK, WARNING, CRITICAL).
#
check_conditions() {
  # check CIRITAL first
  check_condition $2 $3
  RET=$?

  # if CRITICAL
  if [ $RET -eq 1 ]; then
    return $STATE_CRITICAL
  else
    # check WARNING
    check_condition $1 $3
    RET=$?

    # if WARNING
    if [ $RET -eq 1 ]; then
      return $STATE_WARNING
    fi
  fi;

  return $STATE_OK
}

#
# check_api_result()
#
# Parameters:
#  -
#
# Return value:
#  Returns 1 if the result is ok, otherwise returns 0.
#
check_api_result() {
  ERROR=`grep ' ERROR [0-9]*: ' $TEMPFILELOG`
  if [ "$ERROR" != "" ]; then
    echo $ERROR
    return 0
  fi

  RESULT=`grep 'Result=' $TEMPFILE | sed -e 's/^Result=\(.*\)\r$/\1/'`
  if [ "$RESULT" != "OK" ]; then
    INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=\(.*\)\r$/\1/'`
    echo $INFO
    return 0
  fi

  return 1
}

#
# check_exception()
#
# Parameters:
#  1. List of the exceptions, delimited with comma
#  2. Text to check in the list of exceptions
#
# Return value:
#  Returns 1 if the text in the list of exceptions, otherwise return 0.
#
check_exception() {
  local EXCEPTIONS
  local INARRAY

  IFS=',' read -ra EXCEPTIONS <<< "$1"

  INARRAY=$(echo " ${EXCEPTIONS[@]} " | grep -o " $2 " | wc -w)
  if [ $INARRAY -eq 0 ]; then
    return 0
  else
    return 1
  fi
}

#
# main
#
if [ "$1" = "" ]; then
  usage
  exit 0
fi

PROGPATH=$(echo "$0" | sed -e 's,[\\/][^\\/][^\\/]*$,,')
. "$PROGPATH"/utils.sh

CONFIGFILE=/etc/ibminagios.conf

HOST=$1
CMD=$2
TEMPFILE=`mktemp`
TEMPFILELOG=`mktemp`
STATE=$STATE_OK

while read LINE
do
 IP=`echo $LINE | sed -e 's/^\([0-9]*\.[0-9]*\.[0-9]*\.[0-9]*\):AUTHKEY=.*/\1/'`
 if [ "$IP" = "$HOST" ]; then
   AUTHKEY=`echo $LINE | sed -e 's/^[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*:AUTHKEY=\(.*\)$/\1/'`
 fi
 IP=`echo $LINE | sed -e 's/^\([0-9]*\.[0-9]*\.[0-9]*\.[0-9]*\):HTTPPORT=.*/\1/'`
 if [ "$IP" = "$HOST" ]; then
   HTTPPORT=`echo $LINE | sed -e 's/^[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*:HTTPPORT=\(.*\)$/\1/'`
 fi
 if [ "$HTTPPORT" = "" ]; then
   HTTPPORT=80
 fi
done < $CONFIGFILE

URL="http://$HOST:$HTTPPORT/ibminagios?authkey=$AUTHKEY"

case $CMD in

  syssts)
    URL="$URL&cmd=001"
    wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
    check_api_result
    if [ $? -eq 0 ]; then
      STATE=$STATE_UNKNOWN
    else
      SUBCMD=$3

      case $SUBCMD in

        aspused)
          INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=*.|AspUsed=\([0-9]*\).*$/\1/'`
          if [ "$INFO" = "" ]; then
            STATE=$STATE_UNKNOWN
          else
            check_conditions $4 $5 $INFO
            STATE=$?
            echo "$INFO%"
          fi
          ;;

        *)
          STATE=$STATE_UNKNOWN
          echo "Unknown subcommand $SUBCMD"
          ;;

      esac
    fi
    ;;

  job)
    SUBCMD=$3

    URL="$URL&cmd=003&status=*ACTIVE"
    if [ "$SUBCMD" = "act" ]; then
       URL="$URL&jobname=$6"
    fi

    wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
    check_api_result
    if [ $? -eq 0 ]; then
      exit $STATE_UNKNOWN
    else

      case $SUBCMD in

        msgw)
          C=0
          while read LINE
          do
            INFO=`echo $LINE | grep '^Info=|'`
            if [ "$INFO" != "" ]; then
              ACTIVEJOBSTATUS=`echo $INFO | sed -e 's/^Info=.*|ActiveJobStatus=\(.*\)|SubsystemName=.*$/\1/'`
              if [ "$ACTIVEJOBSTATUS" = "MSGW" ]; then
                JOBNAME=`echo $INFO | sed -e 's/^Info=.*|JobName=\(.*\)|JobUser=.*$/\1/'`
                if [ "$6" = "" ]; then
                  let C=C+1
                else
                  check_exception $6 $JOBNAME
                  if [ $? -eq 0 ]; then
                    let C=C+1
                  fi
                fi
              fi
            fi
          done < $TEMPFILE
          check_conditions $4 $5 $C
          STATE=$?
          echo -n $C
          if [ "$6" != "" ]; then
            echo ", exceptions: $6"
          else
            echo ""
          fi
          ;;

        act)
          INFO=`grep '^Info=|' $TEMPFILE | wc -l`
          check_conditions $4 $5 $INFO
          STATE=$?
          echo "$INFO"
          ;;

        *)
          STATE=$STATE_UNKNOWN
          echo "Unknown subcommand $SUBCMD"
          ;;

      esac
    fi
    ;;

  outq)
    URL="$URL&cmd=002&outqname=$6"
    wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
    check_api_result
    if [ $? -eq 0 ]; then
      STATE=$STATE_UNKNOWN
    else
      SUBCMD=$3
      case $SUBCMD in

        splfcount)
          INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=.*|NumberOfFiles=\([0-9]*\).*$/\1/'`
          if [ "$INFO" = "" ]; then
            STATE=$STATE_UNKNOWN
          else
            check_conditions $4 $5 $INFO
            STATE=$?
            echo "$INFO"
          fi
          ;;

        *)
          STATE=$STATE_UNKNOWN
          echo "Unknown subcommand $SUBCMD"
          ;;

      esac
    fi
    ;;

  sql)
    URL="$URL&cmd=004&sql=$6"
    wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
    check_api_result
    if [ $? -eq 0 ]; then
      STATE=$STATE_UNKNOWN
    else
      INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=.*|Value=\(.*\)|.*$/\1/'`
      check_conditions $4 $5 $INFO
      STATE=$?
      echo "$INFO"
    fi
    ;;

  sbs)
    URL="$URL&cmd=005&sbsname=$6"
    wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
    check_api_result
    if [ $? -eq 0 ]; then
      STATE=$STATE_UNKNOWN
    else
      SUBCMD=$3
      case $SUBCMD in

        act)
          INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=.*|Status=\(.*\)|CurrentlyActiveJobs.*$/\1/'`
          check_conditions $4 $5 $INFO
          STATE=$?
          echo "$INFO"
          ;;

        jobcount)
          INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=.*|CurrentlyActiveJobs=\([0-9]*\).*$/\1/'`
          if [ "$INFO" = "" ]; then
            STATE=$STATE_UNKNOWN
            echo "No Information"
          else
            check_conditions $4 $5 $INFO
            STATE=$?
            echo "$INFO"
          fi
          ;;

        *)
          STATE=$STATE_UNKNOWN
          echo "Unknown subcommand $SUBCMD"
          ;;

      esac
    fi
    ;;

  msgq)
    URL="$URL&cmd=006&msgqname=$6&msgid=$7&minutes=$8"
    SUBCMD=$3

    case $SUBCMD in

      inq)
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          INFO=`grep '^Info=.*|Type=05|.*' $TEMPFILE | wc -l`
          check_conditions $4 $5 $INFO
          STATE=$?
          echo "$INFO"
        fi
        ;;

      search)
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          C=0
          while read LINE
          do
            INFO=`echo $LINE | grep '^Info=|'`
            if [ "$INFO" != "" ]; then
              MESSAGEID=`echo $INFO | sed -e 's/^Info=.*|MessageId=\(.*\)|Date=.*$/\1/'`
              check_exception $9 $MESSAGEID
              if [ $? -eq 0 ]; then
                let C=C+1
              fi
            fi
          done < $TEMPFILE
          check_conditions $4 $5 $C
          STATE=$?
          echo -n $C
          if [ "$9" != "" ]; then
            echo ", exceptions: $9"
          else
            echo ""
          fi
        fi
        ;;

      *)
        STATE=$STATE_UNKNOWN
        echo "Unknown subcommand $SUBCMD"
        ;;

    esac
    ;;

  ifs)
    URL="$URL&cmd=007&dir=$6"
    SUBCMD=$3

    case $SUBCMD in

      filecount)
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          INFO=`grep '^Info=.*|Type=FILE|.*$' $TEMPFILE | wc -l`
          check_conditions $4 $5 $INFO
          STATE=$?
          echo "$INFO"
        fi
        ;;

      dircount)
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          INFO=`grep '^Info=.*|Type=DIR|.*$' $TEMPFILE | wc -l`
          check_conditions $4 $5 $INFO
          STATE=$?
          echo "$INFO"
        fi
        ;;

      filesize)
        URL="$URL&file=$7"
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          INFO=`grep 'Info' $TEMPFILE | sed -e 's/^Info=.*|Size=\([0-9]*\).*$/\1/'`
          if [ "$INFO" = "" ]; then
            STATE=$STATE_UNKNOWN
          else
            check_conditions $4 $5 $INFO
            STATE=$?
            echo "$INFO bytes"
          fi
        fi
        ;;

      oldestfile)
        URL="$URL&file=$7"
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          AGE=0
          DATETIME=0
          FILE=""
          while read LINE
          do
            INFO=`echo $LINE | grep '^Info=|.*|Type=FILE|.*'`
            if [ "$INFO" != "" ]; then
              MODIFIEDDATETIME=`echo $INFO | sed -e 's/^Info=.*|ModifiedDateTime=\([0-9]*\).*$/\1/'`
              YYYYMMDD=${MODIFIEDDATETIME:0:8}
              HH=10#${MODIFIEDDATETIME:8:2}
              MM=10#${MODIFIEDDATETIME:10:2}
              SS=10#${MODIFIEDDATETIME:12:2}
              let T=($(date +%s)-$(date +%s -d $YYYYMMDD)-$HH*3600-$MM*60-$SS)
              if [ $T -gt $AGE ]; then
                AGE=$T
                let DATETIME=$(date +%s -d $YYYYMMDD)+$HH*3600+$MM*60+$SS
                FILE=`echo $INFO | sed -e 's/^Info=.*|Name=\(.*\)|Type=.*$/\1/'`
              fi
            fi
          done < $TEMPFILE
          if [ $AGE -gt 0 ]; then
            INFO=$AGE
            check_conditions $4 $5 $INFO
            STATE=$?
            INFO=`date --date=@$DATETIME +"%Y.%m.%d %H:%M:%S"`
            echo "$INFO $FILE"
          else
            echo "-"
          fi 
        fi
        ;;

      emptyfilecount)
        URL="$URL&file=$7"
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          CNT=0
          while read LINE
          do
            INFO=`echo $LINE | grep '^Info=|.*|Type=FILE|.*'`
            if [ "$INFO" != "" ]; then
              SIZE=`echo $INFO | sed -e 's/^Info=.*|Size=\([0-9]*\).*$/\1/'`
              if [ $SIZE -eq 0 ]; then
                let CNT=$CNT+1
              fi
            fi
          done < $TEMPFILE
          INFO=$CNT
          check_conditions $4 $5 $CNT
          STATE=$?
          echo $CNT
        fi
        ;;


      *)
        STATE=$STATE_UNKNOWN
        echo "Unknown subcommand $SUBCMD"
        ;;

    esac
    ;;

  jrnrcv)
    URL="$URL&cmd=008"
    SUBCMD=$3

    case $SUBCMD in

      partialcount)
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          INFO=`grep '^Info=.*|Status=5|.*$' $TEMPFILE | wc -l`
          check_conditions $4 $5 $INFO
          STATE=$?
          if [ $INFO -gt 0 ]; then
            INFO2=`grep '^Info=.*|Status=5|.*$' $TEMPFILE | sed -e 's/^.*|JrnRcvName=\(.*\)|JrnRcvLibrary.*$/\1/'`
            INFO="$INFO, $INFO2"
          fi
          echo "$INFO"
        fi
        ;;

      oldestdetached)
        wget -O $TEMPFILE -o $TEMPFILELOG "$URL"
        check_api_result
        if [ $? -eq 0 ]; then
          STATE=$STATE_UNKNOWN
        else
          if [ -z "$6" ]; then
            JRNLIBRARY=""
            JRNNAME=""
          else
            if [ `echo $6 | grep -l '/' | wc -l` -eq 0 ]; then
              JRNLIBRARY=""
              JRNNAME=$6
            else
              JRNLIBRARY=`echo "$6" | sed -e 's/^\(.*\)[/].*$/\1/'`
              JRNNAME=`echo "$6" | sed -e 's/^.*[/]\(.*\)$/\1/'`
            fi
          fi
          AGE=0
          DATETIME=0
          FILE=""
          LIBRARY=""
          while read LINE
          do
            if [ -z "$JRNLIBRARY" ] && [ -z "$JRNNAME" ]; then
              INFO=`echo $LINE | grep '^Info=|.*|Status=[234]|.*'`
            elif [ -z "$JRNLIBRARY" ] && [ -n "$JRNNAME" ]; then
              C="^Info=|.*|JrnName=$JRNNAME|.*|Status=[234]|.*"
              INFO=`echo $LINE | grep "$C"`
            else
              C="^Info=|.*|JrnName=$JRNNAME|JrnLibrary=$JRNLIBRARY|Status=[234]|.*"
              INFO=`echo $LINE | grep "$C"`
            fi
            if [ "$INFO" != "" ]; then
              MODIFIEDDATETIME=`echo $INFO | sed -e 's/^Info=.*|DetachedDateTime=\([0-9]*\).*$/\1/'`
              YYYYMMDD=${MODIFIEDDATETIME:0:8}
              HH=${MODIFIEDDATETIME:8:2}
              MM=${MODIFIEDDATETIME:10:2}
              SS=${MODIFIEDDATETIME:12:2}
              let T=($(date +%s)-$(date +%s -d $YYYYMMDD)-$HH*3600-$MM*60-$SS)
              if [ $T -gt $AGE ]; then
                AGE=$T
                let DATETIME=$(date +%s -d $YYYYMMDD)+$HH*3600+$MM*60+$SS
                FILE=`echo $INFO | sed -e 's/^Info=.*|JrnRcvName=\(.*\)|JrnRcvLibrary=.*$/\1/'`
                LIBRARY=`echo $INFO | sed -e 's/^Info=.*|JrnRcvLibrary=\(.*\)|JrnName=.*$/\1/'`
              fi
            fi
          done < $TEMPFILE
          if [ $AGE -gt 0 ]; then
            INFO=$AGE
            check_conditions $4 $5 $INFO
            STATE=$?
            INFO=`date --date=@$DATETIME +"%Y.%m.%d %H:%M:%S"`
            echo "$INFO $LIBRARY/$FILE"
          else
            echo "-"
          fi
        fi
        ;;

      *)
        STATE=$STATE_UNKNOWN
        echo "Unknown subcommand $SUBCMD"
        ;;

    esac
    ;;

  *)
    STATE=$STATE_UNKNOWN
    echo "Unknown command $CMD"
    ;;

esac

if [ -e $TEMPFILE ]; then
  rm $TEMPFILE
fi

if [ -e $TEMPFILELOG ]; then
  rm $TEMPFILELOG
fi

exit $STATE
